cpp/src/arrow/io/CMakeLists.txt

```diff
@@ -20,6 +20,7 @@
 
 set(ARROW_IO_LINK_LIBS
   arrow_shared
+  dl
```

Comment:
We are using `dlopen` in `libhdfs_shim.cc`, I get a linker error without linking `libdl`